processing/src/main/java/com/dataspark/masking/MaskingUtil.java

```diff
@@ -0,0 +1,159 @@
+  public static long extrapolate(double value)
+  {
+    return extrapolate(Math.round(value));
+  }
+
+  public static long extrapolate(float value)
+  {
+    return extrapolate(Math.round(value));
+  }
+
+  public static long extrapolate(long value)
+  {
+    return Math.round(value * getExtrapolationFactor());
+  }
+
+  //privacy function
+  public static long applyPrivacyCensor(long value)
+  {
+    final long privacyThreshold = getPrivacyThreshold();
+    return value < privacyThreshold ? privacyThreshold : value;
+  }
+
+  //masking functions
+  public static boolean shouldMask(DataSource dataSource, String fieldName)
+  {
+    if(!maskingOn() || dataSource == null){
+      System.out.println("[ MASKING_UTIL ] will not mask: " + fieldName +" of datasource: " + dataSource);
+      return false;
+    }
+
+    //only do if the datasource is a table ie. true source of data and not a nested query datasource
+    final boolean shouldMask = (dataSource instanceof TableDataSource) && !org.apache.commons.lang.StringUtils.isEmpty(
+        fieldName) && metricsToMask().contains(
+        ((TableDataSource) dataSource).getName().concat(".").concat(fieldName));
+
+    System.out.println("[ MASKING_UTIL ] should mask: " + fieldName +" of datasource: "+ dataSource.getNames() +" = " + shouldMask + " ,DatasourceClass: " + dataSource.getClass());
+    return shouldMask;
+  }
+
+  public static long mask(long value)
+  {
+    return applyPrivacyCensor(extrapolate(value));
+  }
+
+  public static long mask(float value)
+  {
+    return applyPrivacyCensor(extrapolate(value));
+  }
+
+  public static long mask(double value)
+  {
+    return applyPrivacyCensor(extrapolate(value));
+  }
+
+  public static HyperLogLogCollector mask(HyperLogLogCollector hal)
+  {
+    final long privacyThreshold = getPrivacyThreshold();
+    final double originalEstimate = hal.estimateCardinality();
+    final long extrapolatedEstimate = extrapolate(originalEstimate);
+
+    if(extrapolatedEstimate < privacyThreshold) {
+      return createFakeHyperLog(privacyThreshold);
+    }//else return extrapolated number.
+
+    return createFakeHyperLog(extrapolatedEstimate);
+  }
+
+  public static Object doMask(Object value)
+  {
+    if(value instanceof Long) {
+      return mask((Long)value);
+    }else if(value instanceof Double) {
+      return mask((Double)value);
+    }else if(value instanceof Float) {
+      return mask((Float)value);
+    }else if(value instanceof HyperLogLogCollector) {
+      return mask((HyperLogLogCollector) value);
+    }
+    return value;
+  }
+
+  public static HyperLogLogCollector createFakeHyperLog(long numFake)
+  {
+    HyperLogLogCollector collector = HyperLogLogCollector.makeLatestCollector();
+    try {
+      MessageDigest MD5 = MessageDigest.getInstance("MD5");
+      for(long i = 0 ; i < numFake ; i++) {
+        //create fake people :)
+        final String fakeId = "DatasparkFakeId_" + i;
+        collector.add(MD5.digest(fakeId.getBytes()));
```

Comment:
What is the need of Doing MD5 ?

Reply:
the add method needs a byte[] which is hashed. (public void add(byte[] hashedValue))...